components/cam/src/utils/cam_abortutils.F90

```diff
@@ -1,11 +1,75 @@
 module cam_abortutils
 
-  use shr_sys_mod, only: endrun => shr_sys_abort
+!----------------------------------------------------------------------- 
+! 
+! Purpose: This module provides CAM an interface to the model
+!          stopping mechanisms.
+! 
+!-----------------------------------------------------------------------
 
+!-----------------------------------------------------------------------
+!- use statements ------------------------------------------------------
+!-----------------------------------------------------------------------
+  use shr_kind_mod, only: shr_kind_in
+  use shr_sys_mod,  only: shr_sys_abort
+
+!-----------------------------------------------------------------------
+!- module boilerplate --------------------------------------------------
+!-----------------------------------------------------------------------
   implicit none
   private
   save
 
-  public :: endrun
+!-----------------------------------------------------------------------
+! Public interfaces ----------------------------------------------------
+!-----------------------------------------------------------------------
+  public endrun
+
+!-----------------------------------------------------------------------
+! Subroutines and functions --------------------------------------------
+!-----------------------------------------------------------------------
+contains
+
+!========================================================================
+
+   subroutine endrun(string,rc)
+    !--------------------------------------------------------------------
+    ! 
+    ! Purpose: CAM interface to consistent stopping mechanism
+    ! 
+    ! Method: Calls shr_sys_abort
+    ! 
+    !--------------------------------------------------------------------
+
+    implicit none
+
+    !
+    ! Arguments
+    !
+    character(len=*)    , intent(in), optional :: string ! error message 
+                                                         ! string
+    integer(shr_kind_in), intent(in), optional :: rc     ! error code
+    !
+    ! Local workspace
+    !
+    character(len=*), parameter :: &
+      cam_string = "See atm.log for error description" ! generic string
+    !--------------------------------------------------------------------
+
+    if (present(string)) then
+      if (present(rc)) then
+        call shr_sys_abort(string,rc)
+      else
+        call shr_sys_abort(string)
+      endif
+    else
+      if (present(rc)) then
+        call shr_sys_abort(cam_string,rc)
+      else
+        call shr_sys_abort(cam_string)
+      endif
+    endif
```

Comment:
I like the way you organized the code and that it is very clean! Could we re-organize the above if-statements in the following way? Unless I am missing something, it has the same logic but a little bit of less code (removed `parameter` from `cam_string` definition and rearranged if conditions). Also, code is okay in its current state if you don't feel like modifying it.

```
character(len=*) :: &
      cam_string = "See atm.log for error description" ! generic string
....
....
if (present(string)) cam_string = string

if (present(rc)) then
   call shr_sys_abort(ca_string,rc)
else
   call shr_sys_abort(cam_string)
endif
```
Everything else looks great!

Reply:
Looks correct (and much cleaner - thanks for putting in the effort). I'll give it a try and then push a new commit if it works.

Reply:
If the string passed into the routine is longer than cam_string, will cam_string be lengthened automatically (wouldn't think so) or will the new string be truncated to fit into cam_string?

Since endrun is never called with an rc argument, I could leave this off as well. Since endrun is (now) equivalanced to shr_sys_abort, I added this in, but endrun was never defined with an rc argument in the past.  Any opinion on whether rc should be kept or dropped?

Reply:
Nope:

PGF90-S-0089-Illegal use of symbol, cam_string, with character length = * (/autofs/nccs-svm1_home1/worley/E3SM/worleyph/cam/endrun_mod/E3SM/components/cam/\
src/utils/cam_abortutils.F90)
  0 inform,   0 warnings,   1 severes, 0 fatal for endrun

I can give the string a real length (e.g. shr_kind_cl). That should work. Any opinion about how long the string should be? 

The original implementation doesn't require a cap on the string length. Does this make it a better choice?